pkg/forward/forward.go

```diff
@@ -268,6 +265,32 @@ func (isdf *InterStepDataForward) forwardAChunk(ctx context.Context) {
 			}
 		}
 	}
+
+	if isdf.opts.vertexType == dfv1.VertexTypeSource {
+		if isdf.opts.srcWatermarkPublisher == nil {
+			isdf.opts.logger.Errorw("Failed to publish source watermark because no publisher is assigned")
+			return
+		}
+
+		var writeMessages []*isb.Message
+		var transformedReadMessages []*isb.ReadMessage
+		for _, m := range udfResults {
+			writeMessages = append(writeMessages, m.writeMessages...)
+			for _, message := range m.writeMessages {
+				// we convert each writeMessage to isb.ReadMessage by providing its parent ReadMessage's ReadOffset and Watermark.
+				// which is used later for publishing source watermark.
+				transformedReadMessages = append(transformedReadMessages, message.ToReadMessage(m.readMessage.ReadOffset, m.readMessage.Watermark))
+			}
+		}
+		// Assign isLate and publish source watermark
+		for _, m := range writeMessages {
+			if processorWM.After(m.EventTime) { // Set late data at source level
+				m.IsLate = true
```

Comment:
Please test and make sure this message the same one in `messageToStep`.

Reply:
Thanks! messageToStep is not holding pointers, hence the IsLate assignment doesn't take effect. Will fix. +1

Reply:
Fixed by moving the IsLate assignments to before messageToStep assignments.